<a href="https://colab.research.google.com/github/essanchristian-maker/DI-Bootcamp/blob/master/Exercises_XP_Gold_week2_day_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

EXERCICE 1

In [1]:
# Step 1 : Install Kaggle API
!pip install kaggle

# Step 2 : Upload kaggle.json
from google.colab import files
files.upload()

# Step 3 : Set up credentials
import os
os.makedirs("/root/.kaggle", exist_ok=True)
!mv kaggle.json /root/.kaggle/
!chmod 600 /root/.kaggle/kaggle.json

# Step 4 : Download Boston Housing dataset
!kaggle datasets download -d vikrishnan/boston-house-prices

# Step 5 : Load it
import zipfile
import pandas as pd

with zipfile.ZipFile("boston-house-prices.zip", "r") as z:
    z.extractall(".")

boston_df = pd.read_csv("housing.csv", header=None, sep="\s+")

# Add column names
boston_df.columns = ["CRIM","ZN","INDUS","CHAS","NOX",
                     "RM","AGE","DIS","RAD","TAX",
                     "PTRATIO","B","LSTAT","MEDV"]

boston_df.head()

<>:24: SyntaxWarning: invalid escape sequence '\s'
<>:24: SyntaxWarning: invalid escape sequence '\s'
/tmp/ipykernel_522/1245439091.py:24: SyntaxWarning: invalid escape sequence '\s'
  boston_df = pd.read_csv("housing.csv", header=None, sep="\s+")


Saving kaggle (1).json to kaggle (1).json
mv: cannot stat 'kaggle.json': No such file or directory
chmod: cannot access '/root/.kaggle/kaggle.json': No such file or directory
Dataset URL: https://www.kaggle.com/datasets/vikrishnan/boston-house-prices
License(s): other
100% 12.8k/12.8k [00:00<00:00, 30.1MB/s]



,CRIM,ZN,INDUS,CHAS,NOX,RM,AGE,DIS,RAD,TAX,PTRATIO,B,LSTAT,MEDV
0,0.00632,18.0,2.31,0,0.538,6.575,65.2,4.0900,1,296.0,15.3,396.90,4.98,24.0
1,0.02731,0.0,7.07,0,0.469,6.421,78.9,4.9671,2,242.0,17.8,396.90,9.14,21.6
2,0.02729,0.0,7.07,0,0.469,7.185,61.1,4.9671,2,242.0,17.8,392.83,4.03,34.7
3,0.03237,0.0,2.18,0,0.458,6.998,45.8,6.0622,3,222.0,18.7,394.63,2.94,33.4
4,0.06905,0.0,2.18,0,0.458,7.147,54.2,6.0622,3,222.0,18.7,396.90,5.33,36.2


In [2]:
# Step 6 : Basic analysis on MEDV column (Median House Value)

print("Column : MEDV (Median House Value in $1000s)\n")
print(f"Mean    : {boston_df['MEDV'].mean():.2f}")
print(f"Median  : {boston_df['MEDV'].median():.2f}")
print(f"Std Dev : {boston_df['MEDV'].std():.2f}")

Column : MEDV (Median House Value in $1000s)

Mean    : 22.53
Median  : 21.20
Std Dev : 9.20


### Dataset : Boston Housing
The Boston Housing dataset contains 506 rows and 14 columns
describing housing characteristics in Boston suburbs.

### Analysis on column : MEDV (Median House Value in $1000s)
| Metric | Value |
|---|---|
| Mean | 22.53 |
| Median | 21.20 |
| Std Dev | 9.20 |

The mean and median are close, suggesting a roughly symmetric distribution.
The standard deviation of ~9 indicates moderate variability in house prices.

EXERCICE 2

In [5]:
import pandas as pd
import zipfile
import requests
import io

def load_csv_from_zip_url(url):
    response = requests.get(url)
    z = zipfile.ZipFile(io.BytesIO(response.content))
    csv_name = [f for f in z.namelist() if f.endswith(".csv")][0]
    return pd.read_csv(z.open(csv_name))

In [6]:
# Load Retail dataset
url_retail = "https://github.com/devtlv/Datasets-DA-Bootcamp-2-/raw/refs/heads/main/Week%204%20-%20Data%20Understanding/W4D3%20-%20Importing%20Data,%20Exporting%20D/Retail%20Data%20Analytics%20dataset.zip"

retail_df = load_csv_from_zip_url(url_retail)
retail_df.head()

,Store,Date,Temperature,Fuel_Price,MarkDown1,MarkDown2,MarkDown3,MarkDown4,MarkDown5,CPI,Unemployment,IsHoliday
0,1,05/02/2010,42.31,2.572,NaN,NaN,NaN,NaN,NaN,211.096358,8.106,False
1,1,12/02/2010,38.51,2.548,NaN,NaN,NaN,NaN,NaN,211.242170,8.106,True
2,1,19/02/2010,39.93,2.514,NaN,NaN,NaN,NaN,NaN,211.289143,8.106,False
3,1,26/02/2010,46.63,2.561,NaN,NaN,NaN,NaN,NaN,211.319643,8.106,False
4,1,05/03/2010,46.50,2.625,NaN,NaN,NaN,NaN,NaN,211.350143,8.106,False


In [7]:
# Load Women's Clothing Reviews dataset
url_reviews = "https://github.com/devtlv/Datasets-DA-Bootcamp-2-/raw/refs/heads/main/Week%204%20-%20Data%20Understanding/W4D3%20-%20Importing%20Data,%20Exporting%20D/Women's%20E-Commerce%20Clothing%20Reviews.zip"

reviews_df = load_csv_from_zip_url(url_reviews)
reviews_df.head()

,Unnamed: 0,Clothing ID,Age,Title,Review Text,Rating,Recommended IND,Positive Feedback Count,Division Name,Department Name,Class Name
0,0,767,33,NaN,Absolutely wonderful - silky and sexy and comf...,4,1,0,Initmates,Intimate,Intimates
1,1,1080,34,NaN,Love this dress! it's sooo pretty. i happene...,5,1,4,General,Dresses,Dresses
2,2,1077,60,Some major design flaws,I had such high hopes for this dress and reall...,3,0,0,General,Dresses,Dresses
3,3,1049,50,My favorite buy!,"I love, love, love this jumpsuit. it's fun, fl...",5,1,0,General Petite,Bottoms,Pants
4,4,847,47,Flattering shirt,This shirt is very flattering to all due to th...,5,1,6,General,Tops,Blouses


In [9]:
print(retail_df.columns.tolist())
print(retail_df.head())

['Store', 'Date', 'Temperature', 'Fuel_Price', 'MarkDown1', 'MarkDown2', 'MarkDown3', 'MarkDown4', 'MarkDown5', 'CPI', 'Unemployment', 'IsHoliday']
   Store        Date  Temperature  Fuel_Price  MarkDown1  MarkDown2  \
0      1  05/02/2010        42.31       2.572        NaN        NaN   
1      1  12/02/2010        38.51       2.548        NaN        NaN   
2      1  19/02/2010        39.93       2.514        NaN        NaN   
3      1  26/02/2010        46.63       2.561        NaN        NaN   
4      1  05/03/2010        46.50       2.625        NaN        NaN   

   MarkDown3  MarkDown4  MarkDown5         CPI  Unemployment  IsHoliday  
0        NaN        NaN        NaN  211.096358         8.106      False  
1        NaN        NaN        NaN  211.242170         8.106       True  
2        NaN        NaN        NaN  211.289143         8.106      False  
3        NaN        NaN        NaN  211.319643         8.106      False  
4        NaN        NaN        NaN  211.350143         

In [10]:
# Sales trends — using MarkDown1 as main numerical indicator
print("=== TEMPERATURE TREND OVER TIME ===")
print(retail_df.groupby("Date")["Temperature"].mean().sort_index().head(10))

# Store performance — by average CPI and Unemployment
print("\n=== STORE PERFORMANCE (Avg Unemployment) ===")
print(retail_df.groupby("Store")["Unemployment"].mean().sort_values())

# Holiday purchasing habits
print("\n=== AVERAGE TEMPERATURE BY HOLIDAY ===")
print(retail_df.groupby("IsHoliday")["Temperature"].mean())

print("\n=== AVERAGE FUEL PRICE BY HOLIDAY ===")
print(retail_df.groupby("IsHoliday")["Fuel_Price"].mean())

=== TEMPERATURE TREND OVER TIME ===
Date
01/02/2013    42.755778
01/03/2013    40.070667
01/04/2011    48.736444
01/06/2012    72.943778
01/07/2011    78.759111
01/10/2010    69.868000
02/03/2012    45.486444
02/04/2010    53.120889
02/07/2010    77.133333
02/09/2011    79.296222
Name: Temperature, dtype: float64

=== STORE PERFORMANCE (Avg Unemployment) ===
Store
23     4.668562
40     4.668562
4      5.647450
9      5.929456
8      5.947349
5      6.163166
16     6.335308
17     6.365296
6      6.412568
44     6.475822
13     6.762024
41     6.807852
11     7.006006
3      7.006006
25     7.368166
20     7.368166
21     7.403959
2      7.403959
31     7.403959
30     7.403959
1      7.440994
36     7.617491
39     7.617491
37     7.617491
26     7.749970
22     7.963166
27     7.990077
15     7.994580
19     7.994580
42     8.137373
10     8.137373
33     8.269621
32     8.378556
7      8.378556
24     8.481408
45     8.640467
14     8.640467
18     8.712793
35     8.757343
29     9.


### Structured Data : Retail Dataset
Columns : Store, Date, Temperature, Fuel_Price, MarkDown1-5, CPI, Unemployment, IsHoliday

- **Trend analysis** : Temperature and Fuel_Price tracked over time by Date.
- **Store performance** : Unemployment rate compared across stores.
- **Customer habits** : Holiday vs non-holiday periods compared via
  Temperature and Fuel_Price averages.
- This data is clean, numerical and easy to aggregate.

### Challenges
- No direct Sales column — insights are indirect (economic indicators).
- MarkDown columns contain many NaN values, limiting their use.

EXERCICE 3

In [12]:
def load_csv_latin(url):
    response = requests.get(url)
    z = zipfile.ZipFile(io.BytesIO(response.content))
    csv_name = [f for f in z.namelist() if f.endswith(".csv")][0]
    return pd.read_csv(z.open(csv_name), encoding="latin-1")

url_ecommerce = "https://github.com/devtlv/Datasets-DA-Bootcamp-2-/raw/refs/heads/main/Week%204%20-%20Data%20Understanding/W4D3%20-%20Importing%20Data,%20Exporting%20D/E-Commerce%20Dataset.zip"

ecommerce_df = load_csv_latin(url_ecommerce)

print("=== FIRST ROWS ===")
print(ecommerce_df.head())

print("\n=== BASIC INFO ===")
print(f"Rows    : {ecommerce_df.shape[0]}")
print(f"Columns : {ecommerce_df.shape[1]}")
print(f"Column names : {ecommerce_df.columns.tolist()}")
print(ecommerce_df.info())

=== FIRST ROWS ===
  InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom  

=== BASIC INFO ===
Rows    : 541909
Columns : 8
Column names : ['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate', 'UnitPrice', 'CustomerID', 'Country']
<class 'pandas.core.fra

In [13]:
structured = ecommerce_df.select_dtypes(include=['float64', 'int64', 'datetime64']).columns.tolist()
categorical = ecommerce_df.select_dtypes(include=['object']).columns.tolist()

print("Structured (numerical/date) columns :", structured)
print("Categorical columns                 :", categorical)

Structured (numerical/date) columns : ['Quantity', 'UnitPrice', 'CustomerID']
Categorical columns                 : ['InvoiceNo', 'StockCode', 'Description', 'InvoiceDate', 'Country']



### Structured Data Columns
| Column | Type | Reasoning |
|---|---|---|
| Quantity | Numerical | Number of units purchased per transaction |
| UnitPrice | Numerical | Price per unit in £ |
| CustomerID | Numerical | Unique customer identifier |
| InvoiceDate | Date | Should be converted to datetime for trend analysis |

### Categorical Columns (also structured)
| Column | Type | Reasoning |
|---|---|---|
| InvoiceNo | Categorical | Unique transaction identifier |
| StockCode | Categorical | Unique product code |
| Description | Categorical | Product name — borders on unstructured |
| Country | Categorical | Customer's country |

### Suggested Unstructured Data to Complement this Dataset
| Unstructured Data | How it complements structured data |
|---|---|
| Customer reviews | Reveal satisfaction beyond Quantity and UnitPrice |
| Product descriptions | Enrich StockCode and Description for better categorization |
| Customer support chats | Identify issues linked to specific InvoiceNo or StockCode |
| Social media comments | Capture sentiment trends by Country |

### How Unstructured Data Would Be Used
- **Reviews + Quantity** : high sales volume but negative reviews
  signals a quality problem.
- **Product descriptions + StockCode** : auto-classify products
  and improve recommendations.
- **Support chats + Country** : detect regional issues
  and adapt customer service accordingly.

Combining both types gives a 360° view of customer behavior
that structured data alone cannot provide.

EXERCICE  4

In [14]:
url_metro = "https://github.com/devtlv/Datasets-DA-Bootcamp-2-/raw/refs/heads/main/Week%204%20-%20Data%20Understanding/W4D3%20-%20Importing%20Data,%20Exporting%20D/Metro%20Interstate%20Traffic%20Volume.zip"

metro_df = load_csv_from_zip_url(url_metro)

print("=== FIRST ROWS ===")
print(metro_df.head())

print("\n=== BASIC INFO ===")
print(f"Rows    : {metro_df.shape[0]}")
print(f"Columns : {metro_df.shape[1]}")
print(metro_df.info())

=== FIRST ROWS ===
  holiday    temp  rain_1h  snow_1h  clouds_all weather_main  \
0     NaN  288.28      0.0      0.0          40       Clouds   
1     NaN  289.36      0.0      0.0          75       Clouds   
2     NaN  289.58      0.0      0.0          90       Clouds   
3     NaN  290.13      0.0      0.0          90       Clouds   
4     NaN  291.14      0.0      0.0          75       Clouds   

  weather_description            date_time  traffic_volume  
0    scattered clouds  2012-10-02 09:00:00            5545  
1       broken clouds  2012-10-02 10:00:00            4516  
2     overcast clouds  2012-10-02 11:00:00            4767  
3     overcast clouds  2012-10-02 12:00:00            5026  
4       broken clouds  2012-10-02 13:00:00            4918  

=== BASIC INFO ===
Rows    : 48204
Columns : 9
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48204 entries, 0 to 48203
Data columns (total 9 columns):
 #   Column               Non-Null Count  Dtype  
---  ------             

In [15]:
structured = metro_df.select_dtypes(include=['float64', 'int64']).columns.tolist()
categorical = metro_df.select_dtypes(include=['object']).columns.tolist()

print("Structured (numerical) columns :", structured)
print("Categorical columns            :", categorical)

Structured (numerical) columns : ['temp', 'rain_1h', 'snow_1h', 'clouds_all', 'traffic_volume']
Categorical columns            : ['holiday', 'weather_main', 'weather_description', 'date_time']


### Structured Elements
| Column | Type | Reasoning |
|---|---|---|
| traffic_volume | Numerical | Countable integer — directly measurable |
| temp | Numerical | Temperature in Kelvin — continuous measure |
| rain_1h | Numerical | Rainfall in mm — continuous measure |
| snow_1h | Numerical | Snowfall in mm — continuous measure |
| clouds_all | Numerical | Cloud coverage in % — continuous measure |
| date_time | Date/Time | Structured timestamp — can be parsed and indexed |

### Unstructured Elements
| Column | Type | Reasoning |
|---|---|---|
| weather_main | Categorical | Fixed category (Rain, Snow, Clear...) — semi-structured |
| weather_description | Text | Free-form natural language description — unstructured |
| holiday | Text | Named holiday or None — borders on unstructured |

### Why This Categorization ?
- **Structured** : numerical and datetime columns have a fixed format,
  can be directly computed, aggregated and visualized.
- **weather_description** is unstructured because it contains
  free natural language text ("light intensity drizzle",
  "sky is clear") that requires NLP to analyze systematically.
- **holiday** is semi-structured : it is either None or a named
  string — it could be encoded as a binary column (0/1).
- **weather_main** is semi-structured : it looks like a fixed
  category but is stored as free text without strict encoding.

EXERCICE  5

In [17]:
url_movies = "https://github.com/devtlv/Datasets-DA-Bootcamp-2-/raw/refs/heads/main/Week%204%20-%20Data%20Understanding/W4D3%20-%20Importing%20Data,%20Exporting%20D/MovieLens%2020M%20Dataset.zip"

# Correct path with folder prefix and correct filename
ratings_df = pd.read_csv(z.open("MovieLens 20M Dataset/rating.csv"))

print("=== FIRST ROWS ===")
print(ratings_df.head())

print("\n=== COLUMN TYPES ===")
print(ratings_df.dtypes)

print(f"\nRows    : {ratings_df.shape[0]}")
print(f"Columns : {ratings_df.shape[1]}")

=== FIRST ROWS ===
   userId  movieId  rating            timestamp
0       1        2     3.5  2005-04-02 23:53:47
1       1       29     3.5  2005-04-02 23:31:16
2       1       32     3.5  2005-04-02 23:33:39
3       1       47     3.5  2005-04-02 23:32:07
4       1       50     3.5  2005-04-02 23:29:40

=== COLUMN TYPES ===
userId         int64
movieId        int64
rating       float64
timestamp     object
dtype: object

Rows    : 20000263
Columns : 4



### Structured Elements in rating.csv
| Column | Type | Reasoning |
|---|---|---|
| userId | Numerical (int) | Unique integer identifier per user |
| movieId | Numerical (int) | Unique integer identifier per movie |
| rating | Numerical (float) | Score on a fixed scale (0.5 to 5.0) |
| timestamp | Numerical (int) | Unix timestamp — convertible to datetime |

### Why is rating.csv Considered Structured Data ?
- Every column has a **fixed, well-defined format** (integer, float).
- Data is organized in a **tabular structure** with rows and columns.
- Values follow **strict constraints** : ratings are always
  between 0.5 and 5.0, userId and movieId are always integers.
- The dataset can be directly queried, aggregated and analyzed
  without any preprocessing or NLP.
- Timestamps are numerical and can be converted to datetime
  for trend analysis.

This is a perfect example of purely structured data :
every value has a predictable type, range and meaning.

EXERCICE 6

In [18]:
# Step 1 : Install and import Faker
!pip install faker

from faker import Faker
import pandas as pd
import random

fake = Faker()

# Step 2 : Generate 500 products
products = []

for i in range(1, 501):
    products.append({
        "ProductID"   : i,
        "Name"        : fake.catch_phrase(),
        "Description" : fake.text(max_nb_chars=100),
        "Price"       : round(random.uniform(5.99, 999.99), 2)
    })

# Step 3 : Create DataFrame
products_df = pd.DataFrame(products)

print(f"Shape : {products_df.shape}")
products_df.head()

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.0/2.0 MB 19.6 MB/s eta 0:00:00
Shape : (500, 4)


,ProductID,Name,Description,Price
0,1,Decentralized high-level knowledge user,Leader surface stand set wind build speech. Ev...,185.54
1,2,Automated impactful capacity,Degree role determine money pattern small. Est...,23.92
2,3,User-friendly directional workforce,Civil mother offer toward of feel. Control pre...,445.12
3,4,Horizontal fault-tolerant architecture,Available ground adult beautiful interest need...,638.33
4,5,Operative empowering emulation,Item create cell. Maintain vote worker method ...,381.79


 **Faker** generates realistic fake text data (names, descriptions).
- **random.uniform(5.99, 999.99)** produces realistic price ranges.
- **ProductID** is a unique sequential integer from 1 to 500.
- The final DataFrame has 500 rows and 4 columns :
  ProductID, Name, Description, Price.